### PR TITLE
Swift Archival & Serialization API

### PR DESCRIPTION
**Pull request by itaiferber**
**What's in this pull request?**
Now that the core types introduced by [SE-0166](https://github.com/apple/swift-evolution/blob/master/proposals/0166-swift-archival-serialization.md) have been sunk into the Swift standard library, the stdlib and compiler portions of #8124 and #8125 can be combined into one PR (this one).

The features can be tested together; #8125 was previously not tested on its own since it required the presence of #8124.

The Foundation portions of #8124 will be put up in a subsequent PR.

rdar://problem/30472044

### CODE REVIEW

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
I'm not sure these test failures are actually related to my changes. On Linux we're getting failures like `error: 'Iterator' is not a member type of 'Sequence'`, and on macOS, the failing tests are SourceKit. The Linux failures are more worrisome to me though; anyone know what's going on there?

**Comment by itaiferber** [comment]
Let's try again

**Comment by itaiferber** [comment]
@swift-ci Please test

File: stdlib/public/core/Codable.swift [hunk 1]
```diff
@@ -0,0 +1,3140 @@
...
+
+    /// Decodes a value of the given type for the given key.
+    ///
+    /// - parameter type: The type of value to decode.
+    /// - parameter key: The key that the decoded value is associated with.
+    /// - returns: A value of the requested type, if present for the given key and convertible to the requested type.
+    /// - throws: `DecodingError.typeMismatch` if the encountered encoded value is not convertible to the requested type.
+    /// - throws: `DecodingError.keyNotFound` if `self` does not have an entry for the given key.
+    /// - throws: `DecodingError.valueNotFound` if `self` has a null entry for the given key.
+    func decode<T : Decodable>(_ type: T.Type, forKey key: Key) throws -> T
```
Comment (obitow) [inline stdlib/public/core/Codable.swift]: @itaiferber great proposal, can't wait to start using this. But I have a question: wouldn't be better to drop the type from the parameters and just use inference here? Something like:
```
func decode<T : Decodable>(_ key: Key) throws -> T
```

Comment (itaiferber) [inline stdlib/public/core/Codable.swift]: @obitow Thanks!

This point has been discussed at length on swift-evolution; dropping the type parameter can lead to ambiguous statements, which can be confusing and unintuitive. Please feel free to refer to those discussions for the full story.

**Comment by itaiferber** [comment]
@atrick @gottesmm Looks like the tests are failing in SIL validation on Linux and the errors are kind of bizarre: e.g. `error: 'Iterator' is not a member type of 'Sequence'`. Any idea about what's going on here/how I can address this?

**Comment by itaiferber** [comment]
Let's try the smoke tests at least?

**Comment by itaiferber** [comment]
Perhaps we can try the smoke tests at least?

**Comment by itaiferber** [comment]
@swift-ci Please smoke test

**Comment by itaiferber** [comment]
It's possible that `Sequence : Swift.Sequence` was causing ambiguity when generating SIL. I've changed that; let's see if that's caused any changes in the tests. Once more with vigor...

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
Looks like the `Sequence` change did fix the SIL tests. The SourceKit tests are still failing though.

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
`bdfa927` should have hopefully fixed the SourceKit and `swift-ide-test` issues as `ASTPrinter` was making assumptions about extensions that `typealias Codable = Encodable & Decodable` broke

**Comment by itaiferber** [comment]
Looks like the macOS build failure is not unique to this PR. (See also https://ci.swift.org/job/oss-swift_tools-RA_stdlib-RDA_test-simulator/3143/consoleFull)
Waiting for a resolution on the issue and I'll re-test on macOS.

**Comment by itaiferber** [comment]
Waiting on #9081 to go in before we can retest

**Comment by itaiferber** [comment]
Should be good to go, hopefully

**Comment by itaiferber** [comment]
@swift-ci Please test

File: stdlib/public/core/Codable.swift [hunk 2]
```diff
@@ -0,0 +1,3140 @@
...
+
+/// Conformance to `Encodable` indicates that a type can encode itself to an external representation.
+public protocol Encodable {
+    /// Encodes `self` into the given encoder.
+    ///
+    /// If `self` fails to encode anything, `encoder` will encode an empty keyed container in its place.
+    ///
+    /// - parameter encoder: The encoder to write data to.
+    /// - throws: An error if any values are invalid for `encoder`'s format.
+    func encode(to encoder: Encoder) throws
```
Comment (rintaro) [inline stdlib/public/core/Codable.swift]: Out of curiosity:
Why this is not `func encode<E : Encoder>(to encoder: E) throws`?
I believe generics have more optimization opportunities than protocols existentials.

**Comment by itaiferber** [comment]
I'm wondering if this macOS failure is extant, or actually related to any of my stuff.

**Comment by itaiferber** [comment]
@swift-ci Please test macOS

**Comment by DougGregor** [comment]
It looks like we're getting some duplicate declarations somehow in that SourceKit failure:

    23:37:31              key.kind: source.lang.swift.ref.function.constructor,
    23:37:31              key.name: "init(rawValue:)",
    23:37:31              key.usr: "s:s16RawRepresentablePxSg0A5ValueQz03rawC0_tcfc"
    23:37:31 +          },
    23:37:31 +          {
    23:37:31 +            key.kind: source.lang.swift.ref.function.constructor,
    23:37:31 +            key.name: "init(rawValue:)",
    23:37:31 +            key.usr: "s:s16RawRepresentablePxSg0A5ValueQz03rawC0_tcfc"
    23:37:31            }

**Comment by itaiferber** [comment]
@DougGregor Yeah, I'm investigating why this happens — seems weird considering my stuff doesn't touch these types at all, or defines anything with `init(rawValue:)`. This reproduces locally with my changes, so it's definitely a present issue; trying to see if I can reproduce it without my changes, and in parallel, see if I can track down what about SourceKit is causing this to happen. `sourcekitd-test` connects to SourceKit over XPC, so I want to find out what instance of the daemon it's connecting to; a mismatch may cause weird behavior...

**Comment by itaiferber** [comment]
Not sure how SourceKit could see two `init(rawValue:)` constructors to begin with, since that shouldn't be possible to declare.

**Comment by itaiferber** [comment]
It looks like this change is caused by the inclusion of `extension RawRepresentable where RawValue == ..., Self : Codable { ... }`; commenting out those extensions prevents the repeated conformance.

That being said, I don't know, then, if this SourceKit output is expected (in which case I will update the test to match), or if this is a SourceKit bug, or otherwise. @DougGregor Any ideas? Who can I consult with about this?

**Comment by DougGregor** [comment]
Weird, because we're seeing two entities with exactly the same USR (`s16RawRepresentablePxSg0A5ValueQz03rawC0_tcfc`), which is really odd. Perhaps a consequence of projecting the protocol extensions onto the concrete type. @akyrtzi or @nkcsgexi , any ideas?

**Comment by DougGregor** [comment]
It seems like a bug like this in the actual AST (e.g., duplicated initializers) would cause everything to break horribly, so it feels like a SourceKit bug.

**Comment by itaiferber** [comment]
This last commit doesn't fix the issue, but if I'm already looking at the extensions... (Let's split them up so you can benefit even by just being `Encodable` or `Decodable`.)

File: stdlib/public/core/Codable.swift [hunk 3]
```diff
@@ -1833,6 +1833,14 @@ public struct CodingUserInfoKey : RawRepresentable, Equatable, Hashable {
         self.rawValue = rawValue
     }
 
+    /// Returns whether the given keys are equal.
+    ///
+    /// - parameter lhs: The key to compare against.
+    /// - parameter rhs: The key to compare with.
+    public static func ==(_ lhs: CodingUserInfoKey, _ rhs: CodingUserInfoKey) -> Bool {
```
Comment (itaiferber) [inline stdlib/public/core/Codable.swift]: I'd meant to split this out into a different commit, but it made it in here anyway. `CodingUserInfoKey` was missing `==`, but the compiler didn't complain.

**Comment by itaiferber** [comment]
Per @akyrtzi's suggestion via email, going to update the test to prevent this from blocking us and file a Radar about it.

**Comment by itaiferber** [comment]
@swift-ci Please test and merge

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
Build failed with an unrelated issue that's been reverted; let's try again.

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
For interested Apple folks, BTW, rdar://31896858

**Comment by itaiferber** [comment]
Looks like a Migrator test is failing on iOS: `Migrator/wrap_optional.swift`
Looking into it.

**Comment by itaiferber** [comment]
The builder is producing a module file for the incorrect target? `error: module file was created for incompatible target x86_64-apple-macosx10.9`

**Comment by itaiferber** [comment]
@shahmishal @DougGregor This... Seems wrong. Any idea why this might happen?

**Comment by itaiferber** [comment]
Seems like https://ci.swift.org/view/Pull%20Request/job/swift-PR-osx/ failed with this as well...

**Comment by itaiferber** [comment]
Waiting on #9113 before testing again.

**Comment by itaiferber** [comment]
Let's try that again

**Comment by itaiferber** [comment]
@swift-ci Please test

**Comment by itaiferber** [comment]
Excellent. @parkera @DougGregor If either of sees this, can you please merge it in?

File: stdlib/public/core/Codable.swift (hunk 2 above)
Comment (itaiferber) [inline stdlib/public/core/Codable.swift]: @rintaro `encode(to:)` and `init(from:)` are not generic because there are cases where you need to pass an existential `Encoder` or `Decoder` to an object. For instance, given:

```swift
class A : Codable { /* ... */ }
class B : A { /* ... */ }
```

`class B` would likely implement `encode(to:)` as

```swift
func encode(to encoder: Encoder) throws {
    var container = encoder./* some container type */
    // encode properties into container

    super.encode(to: container.superEncoder())
}
```

This is not possible to do if `encode(to:)` is generic, because `superEncoder()` cannot return a concrete type, and `encode<E : Encoder>(to encoder: E)` cannot take an existential type.

**Comment by itaiferber** [comment]
Thanks @DougGregor!

File: stdlib/public/core/Codable.swift (hunk 2 above)
Comment (rintaro) [inline stdlib/public/core/Codable.swift]: Ah, I've forgotten that `<T : P>` can't receive `P`.
Thank you for clarifying!

File: stdlib/public/core/Codable.swift [hunk 4]
```diff
@@ -0,0 +1,3176 @@
...
+    /// Stores a new nested container for the given key and returns a new `Encoder` instance for encoding `super` into that container.
+    ///
+    /// - parameter key: The key to encode `super` for.
+    /// - returns: A new `Encoder` to pass to `super.encode(to:)`.
+    mutating func superEncoder(forKey key: Key) -> Encoder
+}
+
+// An implementation of _KeyedEncodingContainerBase and _KeyedEncodingContainerBox are given at the bottom of this file.
+/// `KeyedEncodingContainer` is a type-erased box for `KeyedEncodingContainerProtocol` types, similar to `AnyCollection` and `AnyHashable`. This is the type which consumers of the API interact with directly.
+public struct KeyedEncodingContainer<K : CodingKey> {
```
Comment (norio-nomura) [inline stdlib/public/core/Codable.swift]: @itaiferber Should `KeyedEncodingContainer` declare` KeyedEncodingContainerProtocol` conformance?

File: stdlib/public/core/Codable.swift [hunk 5]
```diff
@@ -0,0 +1,3176 @@
...
+    /// - parameter key: The key that the decoded value is associated with.
+    /// - returns: A decoded value of the requested type, or `nil` if the `Decoder` does not have an entry associated with the given key, or if the value is a null value.
+    /// - throws: `DecodingError.typeMismatch` if the encountered encoded value is not convertible to the requested type.
+    public func decodeIfPresent(_ type: String.Type, forKey key: Key) throws -> String? {
+        return try box.decodeIfPresent(String.self, forKey: key)
+    }
+
+    /// Decodes a value of the given type for the given key, if present.
+    ///
+    /// This method retu
```
Comment (norio-nomura) [inline stdlib/public/core/Codable.swift]: @itaiferber Why is it `mutating` only for `decode` of `UnkeyedDecodingContainer`?

Comment (itaiferber) [inline stdlib/public/core/Codable.swift]: `UnkeyedDecodingContainer`s need to maintain some form of state to keep track of where they are in decoding, whether an index into an array that's incrementing with every `decode(...)` call, an offset into a data blob or stream, etc. The methods are `mutating` to indicate that, and to allow the implementation to be a `struct`. This is as opposed to `KeyedDecodingContainer`s and `SingleValueDecodingContainer`s, which by design will almost certainly not need to do that. Is there a use case you have in mind for making those `mutating` as well?

File: stdlib/public/core/Codable.swift (hunk 4 above)
Comment (itaiferber) [inline stdlib/public/core/Codable.swift]: It can, but I don't see much of a benefit to doing so. When would you want to use it as `KeyedEncodingContainerProtocol` instead of the fully-typed `KeyedEncodingContainer<K>`?

Comment (norio-nomura) [inline stdlib/public/core/Codable.swift]: Indeed, you are right. I can not think of such a scene.
I just wondered watching `KeyedDecodingContainer` declared as`KeyedDecodingContainerProtocol`.

File: stdlib/public/core/Codable.swift (hunk 5 above)
Comment (norio-nomura) [inline stdlib/public/core/Codable.swift]: Certainly, I could not think of any reason to make `decode()` of another protocol `mutating` as well.
Thank you for answering my boring questions. :pray:

**Comment by norio-nomura** [comment]
I filed [SR-4861](https://bugs.swift.org/browse/SR-4861) that request adding `encode<T>(_:)` and `decode<T>(_:)` to `SingleValueEncodingContainer` and `SingleValueDecodingContainer`.

File: stdlib/public/core/Codable.swift (hunk 4 above)
Comment (itaiferber) [inline stdlib/public/core/Codable.swift]: Good catch — I don't think that's really necessary either. It doesn't hurt, but we should be consistent. I'll consider this; thanks!

File: stdlib/public/core/Codable.swift (hunk 5 above)
Comment (itaiferber) [inline stdlib/public/core/Codable.swift]: Not at all! Thank you for paying such close attention to this; it's the best way for us to get direct feedback on it, and I appreciate your efforts.

**Comment by itaiferber** [comment]
@norio-nomura Thanks, assigned that to myself. That change, among a few others, is planned, and going through some API review soon.

File: test/decl/protocol/special/coding/class_codable_inheritance.swift [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+  // These lines have to be within the SimpleChildClass type because
+  // CodingKeys should be private.
+  func foo() {
+    // They should receive a synthesized CodingKeys enum.
+    // NOTE: This expected error will need to be removed in the future.
+    let _ = SimpleChildClass.CodingKeys.self // expected-error {{'CodingKeys' is inaccessible due to 'private' protection level}}
+
+    // The enum should have a case for each of the vars.
+    // NOTE: This expectedxerror will need to be removed in the future.
```
Comment (Jceciliani) [inline test/decl/protocol/special/coding/class_codable_inheritance.swift]: Small text fix needed in comment. "//NOTE: This expected error will need to be removed in the future."

**Review by Jceciliani — COMMENTED** [review]
All committed changes appear correct after review.

File: test/decl/protocol/special/coding/class_codable_inheritance.swift (hunk 6 above)
Comment (itaiferber) [reply test/decl/protocol/special/coding/class_codable_inheritance.swift]: Thanks! #16749
